src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts

```diff
@@ -286,17 +286,20 @@ export class TerminalTaskSystem extends Disposable implements ITaskSystem {
 	public run(task: Task, resolver: ITaskResolver, trigger: string = Triggers.command): ITaskExecuteResult {
 		task = task.clone(); // A small amount of task state is stored in the task (instance) and tasks passed in to run may have that set already.
 		const recentTaskKey = task.getRecentlyUsedKey() ?? '';
-		const validInstance = task.runOptions && task.runOptions.instanceLimit && this._instances[recentTaskKey] && this._instances[recentTaskKey].instances < task.runOptions.instanceLimit;
+		const validInstance = !this._instances[recentTaskKey] || this._instances[recentTaskKey].instances < ((task.runOptions && task.runOptions.instanceLimit) ?? 0);
 		const instance = this._instances[recentTaskKey] ? this._instances[recentTaskKey].instances : 0;
 		this._currentTask = new VerifiedTask(task, resolver, trigger);
 		if (instance > 0) {
 			task.instance = this._instances[recentTaskKey].counter;
 		}
 		const lastTaskInstance = this.getLastInstance(task);
 		const terminalData = lastTaskInstance ? this._activeTasks[lastTaskInstance.getMapKey()] : undefined;
-		if (terminalData && terminalData.promise && !validInstance) {
+		if (!validInstance) {
```

Comment:
is there a reason we don't want to also check `&& terminalData && terminalData.promise)` here before setting `this._lastTask = this._currentTask`?

Reply:
For one thing, `promise` isn't nullable, so checking `terminalData.promise` was just redundant.

But this was pretty much the point of this change. If we know that we've exceeded the instance count, it doesn't matter whether there's an entry in _activeTasks or not (especially since prior to #180617 there was a race, and _activeTasks doesn't get updated until some time after the task has been launched).

So the changes here are no longer strictly needed (but we do need to undo the newly added `&& !task.configurationProperties.dependsOn?.length`), but I think it's worth keeping anyway. The logic here (maintained in `_instances`) is self contained; we shouldn't ignore it because of some external conditions.

Reply:
Sorry, I'd forgotten exactly what was going on here, and a lot of what I said above is irrelevant to your question.

Yes, setting this._lastTask should probably be moved under the `terminalData && terminalData.promise` check (although, as I said above, the terminalData.promise check is redundant).

However, due to #180617 it really shouldn't be possible for validInstance to be false, and terminalData to be null any more (but I don't think *this* code should rely on that).